nodejs/guestbook/src/frontend/app.js

```diff
@@ -0,0 +1,59 @@
+const express = require('express')
+const path = require('path');
+const moment = require('moment')
+const app = express();
+const bodyParser = require('body-parser')
+const axios = require('axios')
+
+const util = require('./utils')
+
+const GUESTBOOK_API_ADDR = process.env.GUESTBOOK_API_ADDR || 'localhost:8080'
+
+const BACKEND_URI = `http://${GUESTBOOK_API_ADDR}/messages`
+
+app.set("view engine", "pug")
+
+app.set("views", path.join(__dirname, "views"))
+
+const router = express.Router()
+
+app.use(router)
+router.use(bodyParser.urlencoded({ extended: false }))
+
+// starts an http server on the $PORT environment variable
+const PORT = process.env.PORT || 8001;
+app.listen(PORT, () => {
+  console.log(`App listening on port ${PORT}`);
+  console.log('Press Ctrl+C to quit.');
+});
+
+router.get("/", (req, res) => {
+    // retrieve list of messages from the backend, and use them to render the HTML template
+    axios.get(BACKEND_URI)
+      .then(response => {
```

Comment:
Are you running this with Node 8 and newer? Then I'd go with `async/await`. It's much easier to read than promises: 

```
router.get("/", async (req, res) => {
...
  try {
      const response =  await axios.get(BACKEND_URI);
      console.log('got response: ' + response.data)
      const result = util.formatMessages(response.data)
      res.render("home", {messages: result})
    } catch (error) {
        console.log('error in get request: ' + error)
    }
}
```